### PR TITLE
Fix vwap

### LINKED ISSUE #7880: Vwap not update following a trade
Party's vwap of buy/sell orders does not get updated when those orders trade

### PR DESCRIPTION
**Pull request by witgaw**
* Update buy / sell order vwap after a trade
* Don't calculate running average (keep track of sum of price * size for party's orders)
* Use sum product in snapshot (initial step towards making the VWAP a decimal so we get full precision for auction margins)

Fixes https://github.com/vegaprotocol/vega/issues/7880

### CODE REVIEW

File: core/settlement/engine_test.go [hunk 1]
```diff
@@ -830,12 +830,26 @@ func (t testPos) Price() *num.Uint {
 	return t.price
 }
 
+func (t testPos) BuySumProduct() *num.Uint {
+	return num.NewUint(t.sellSumProduct)
```
Comment (ze97286) [inline core/settlement/engine_test.go:834]: ?

File: core/settlement/engine_test.go [hunk 2]
```diff
@@ -830,12 +830,26 @@ func (t testPos) Price() *num.Uint {
...
+}
+
+func (t testPos) SellSumProduct() *num.Uint {
+	return num.NewUint(t.buySumProduct)
```
Comment (ze97286) [inline core/settlement/engine_test.go:838]: ?

**Review by davidsiska-vega — DISMISSED** [review]
Feature tests LGTM

File: core/settlement/engine_test.go (hunk 1 above)
Comment (witgaw) [reply core/settlement/engine_test.go:834]: Good spot, fixed now.

**Review by ze97286 — DISMISSED** [review]

File: core/positions/market_position.go [hunk 3]
```diff
@@ -59,74 +59,54 @@ func (p *MarketPosition) Closed() bool {
...
+		// recalculate sumproduct
+		if add {
+			p.buySumProduct.Add(p.buySumProduct, num.UintZero().Mul(price, num.NewUint(sizeChange)))
+			p.buy += iSizeChange
 		} else {
-			p.vwBuyPrice.SetUint64(0)
+			p.buySumProduct.Sub(p.buySumProduct, num.UintZero().Mul(price, num.NewUint(sizeChange)))
+			p.buy -= iSizeChange
+		}
+		if p.buy == 0 && !p.buySumProduct.IsZero() {
```
[the remark this answers is not in the document]
Comment (witgaw) [reply core/positions/market_position.go:89]: But it's not the same thing, it's vwap without the division by volume.

**Review by EVODelavega — CHANGES_REQUESTED** [review]

**Review by davidsiska-vega — DISMISSED** [review]

File: core/positions/market_position.go [hunk 4]
```diff
@@ -186,18 +166,36 @@ func (p MarketPosition) Price() *num.Uint {
...
 	return num.UintZero()
 }
 
 // VWSell - get volume weighted sell price for unmatched sell orders.
 func (p MarketPosition) VWSell() *num.Uint {
-	if p.vwSellPrice != nil {
-		return p.vwSellPrice.Clone()
+	if p.sellSumProduct != nil && p.sell != 0 {
+		ret := p.sellSumProduct.Clone()
+		return ret.Div(ret, num.NewUint(uint64(p.sell)))
```
[the remark this answers is not in the document]
Comment (witgaw) [reply core/positions/market_position.go]: Fair enough, I got rid of the clone. 

I think the name is fine, it's not really total value, it's derivatives so it's not quite that simple. If you go to the definition you get the comment `// sum of size*price for party's buy/sell orders` which I believe is sufficient.

**Review by davidsiska-vega — APPROVED** [review]

**Review by EVODelavega — APPROVED** [review]

**Pull request merged**